Sources/Plasma/PubUtilLib/plVault/plVaultClientApi.cpp

```diff
@@ -367,57 +324,53 @@ static void BuildNodeTree (
 ) {
     for (unsigned i = 0; i < refCount; ++i) {
         // Find/Create global links
-        RelVaultNodeLink * parentLink = s_nodes.Find(refs[i].parentId);
-        if (!parentLink) {
+        auto parentIt = s_nodes.find(refs[i].parentId);
+        hsRef<RelVaultNode> parentNode;
+        if (parentIt == s_nodes.end()) {
             newNodeIds->emplace_back(refs[i].parentId);
-            parentLink = new RelVaultNodeLink(false, 0, refs[i].parentId);
-            parentLink->node->SetNodeId_NoDirty(refs[i].parentId);
-            s_nodes.Add(parentLink);
-        }
-        else {
+            parentNode.Steal(new RelVaultNode());
+            parentNode->SetNodeId_NoDirty(refs[i].parentId);
+            parentIt = s_nodes.emplace(refs[i].parentId, parentNode).first;
+        } else {
+            parentNode = parentIt->second;
             existingNodeIds->emplace_back(refs[i].parentId);
         }
-        RelVaultNodeLink * childLink = s_nodes.Find(refs[i].childId);
-        if (!childLink) {
+
+        auto childIt = s_nodes.find(refs[i].childId);
+        hsRef<RelVaultNode> childNode;
```

Comment:
I wonder if this should be `hsWeakRef` so we don't have to explicitly steal the new reference in 343, and we don't have to do an atomic operation in the else block.

Reply:
Hmm, will that work out without potentially leaking references? In the `if` branch, something needs to take ownership of the initial reference from the `new RelVaultNode()` call. If I put that directly into a `hsWeakRef`, it will just leak, and if I `UnRef` it first, it deallocates itself before it reaches a strong `hsRef`...

Best solution I can think of would be a temporary strong `hsRef` that holds the `new RelVaultNode()` long enough to be reffed elsewhere, *and* a `hsWeakRef` that holds either the `new RelVaultNode()` or an already existing one. That *should* avoid temporarily increasing the refcount.

Reply:
Oh wait, the `RelVaultNode` gets stored into a strong `hsRef` by the `emplace` call a few lines later, so I can let that steal it and store it as a plain pointer locally.